### PR TITLE
fix nil printer panic in oc set sub-commands

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes https://github.com/openshift/origin/issues/11439

##### Sub-commands that potentially need to be addressed
- [x] `set/env.go`
- [x] `set/buildsecret.go`
- [x] `set/triggers.go`
- [x] `set/volume.go`
- [x] `set/probe.go`
- [x] `set/buildhook.go`
- [x] `set/deploymenthook.go`

@openshift/cli-review @csrwng

### CODE REVIEW

**Comment by csrwng** [comment]
@juanvallejo looks good. Just curious, why don't we support other output formats?

**Comment by juanvallejo** [comment]
@csrwng from what I have seen, `oc set ...` commands seem to support [these output formats](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L71-L128), although with formats such as `custom-columns` which require a `custom-columns=COL1` syntax, the `=` is not handled between key-value pairs, preventing the format from being parsed correctly.

In the case of the `-o wide` print format, I tried to see if I could simply return a `HumanReadablePrinter` [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L138), but after trying this out, I found out that some resources, such as `DeploymentConfigs` won't work, causing an error output such as this one to be printed instead:

```
$  oc new-app node -o yaml | oc set env -f - MYVAR=value -o wide

error: unknown type &v1.DeploymentConfig{TypeMeta:unversioned.TypeMeta{Kind:"DeploymentConfig", APIVersion:"v1"}, ObjectMeta:v1.ObjectMeta{Name:"node", GenerateName:"", Namespace:"default", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"app":"node"}, Annotations:map[string]string{"openshift.io/generated-by":"OpenShiftNewApp"}, OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1.DeploymentConfigSpec{Strategy:v1.DeploymentStrategy{Type:"Rolling", CustomParams:(*v1.CustomDeploymentStrategyParams)(nil), RecreateParams:(*v1.RecreateDeploymentStrategyParams)(nil), RollingParams:(*v1.RollingDeploymentStrategyParams)(0xc42085ea40), Resources:v1.ResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList(nil)}, Labels:map[string]string(nil), Annotations:map[string]string(nil)}, MinReadySeconds:0, Triggers:v1.DeploymentTriggerPolicies{v1.DeploymentTriggerPolicy{Type:"ConfigChange", ImageChangeParams:(*v1.DeploymentTriggerImageChangeParams)(nil)}, v1.DeploymentTriggerPolicy{Type:"ImageChange", ImageChangeParams:(*v1.DeploymentTriggerImageChangeParams)(0xc42087d220)}}, Replicas:1, RevisionHistoryLimit:(*int32)(nil), Test:false, Paused:false, Selector:map[string]string{"app":"node", "deploymentconfig":"node"}, Template:(*v1.PodTemplateSpec)(0xc4204d14a0)}, Status:v1.DeploymentConfigStatus{LatestVersion:0, ObservedGeneration:0, Replicas:0, UpdatedReplicas:0, AvailableReplicas:0, UnavailableReplicas:0, Details:(*v1.DeploymentDetails)(nil), Conditions:[]v1.DeploymentCondition(nil)}}
```

**Comment by csrwng** [comment]
hmm, how do we support it when doing `oc get`?

**Comment by juanvallejo** [comment]
@csrwng in `oc get`, we pass the `Get` command our `cmdutil.Factory`. When `get.go` ends up calling the factory method `PrinterForMapping` on our factory, it [returns a `NewHumanReadablePrinter`](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L488) instantiated in our [`describe` package with handlers for resource types specific to openshift](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/printer.go#L68).

A solution so that we can also support the `wide` output format on `set` commands could be that we detect if the printer returned by [`GetPrinter`](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/set/env.go#L490) is nil and if the format specified was `"wide"`, and in that specific case use a `describe.NewHumanReadablePrinter` to print the objects instead. What do you think?

**Comment by csrwng** [comment]
@juanvallejo just thinking what would be the right thing to do here ... my preference would be to factor out the code that prints objects from the upstream `get` command and call that from wherever we need to print a set of objects in our CLI commands. It looks like a good number of our commands don't implement all the printing options introduced upstream and we could use a simple function to print a set of objects.

**Comment by juanvallejo** [comment]
@csrwng 
>  my preference would be to factor out the code that prints objects from the upstream get command and call that from wherever we need to print a set of objects in our CLI commands

I agree, since all of our commands already receive a pointer to our `cmdutil.Factory`, we could make use of Printer methods there. For cases like this one where we have to choose a printer based on the output format flag, perhaps we could extend the [`cmdutil.Factory` `Printer](https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L488)` method to receive an `outputFormat` parameter that would allow it to decide which printer to return

**Comment by juanvallejo** [comment]
@csrwng Added a new factory method to support more kinds of output formats, including `wide`. What do you think?  https://github.com/openshift/origin/pull/11448/commits/3379317b9ce4a23100636053a62a761d1b709e33

**Comment by csrwng** [comment]
@juanvallejo hmm, not sure that quite does it. The problem is when you're outputting a heterogeneous list of objects, you can't just use the mapping of the first object in the list. You need something like the output of `get all`

**Comment by juanvallejo** [comment]
@csrwng That makes sense, what if I did something like this:
```
for _, info := range infos {
        mapping := info.ResourceMapping(()
        p, _, err := f.PrinterForOutput(cmd, mapping, outputFormat, "", false)
        if err != nil {
		return err
	}
        if err := p.PrintObj(info.Object, out); err != nil {
		return err
	}
}
```

Also, instead of having the factory method `PrinterForOutput`, I could make a wrapper function for `kubectl.GetPrinter` and call it something like `kubectl.GetVersionedPrinter`, which would pretty much do what the new factory method does now. What do you think?

**Comment by juanvallejo** [comment]
> Also, instead of having the factory method PrinterForOutput, I could make a wrapper function for kubectl.GetPrinter and call it something like kubectl.GetVersionedPrinter, which would pretty much do what the new factory method does now

The reason for having this versioned printer function is that I cannot call `PrintObj` on a `HumanReadablePrinter` with a `v1` object, as it would give an error that it does not support the resource type. Having this wrapper function would wrap the `kubectl.GetPrinter` in a `NewVersionedPrinter` for all printer types except the `HumanReadablePrinter`

**Comment by csrwng** [comment]
Actually, I was thinking that this code can be factored out into a separate utility function:
https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/cmd/get.go#L301-L495
and have both the get command and our cli commands just invoke it. It would require an upstream change, but would ensure that we inherit upstream changes to how things are output. 

I haven't looked closely enough to make sure it would actually make sense, so feel free to tell me that I'm crazy :-)

**Comment by juanvallejo** [comment]
@csrwng It definitely makes sense. It looks like `PrinterForMapping` [calls PrinterForCommand](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L1281), and then attempts to [return a versioned printer here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L1303). Also, PrinterForCommand does require all `PrinterFlags` to have been set, but it also [makes a call to `kubectl.GetPrinter`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/printing.go#L128).

So maybe instead of a new function `PrinterForOutput`, we could instead call `PrinterForMapping`, and slightly [modify this line](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go#L1299) to:
```go
if ok && printer != nil {
        ...
}
```
With this change, it will make the call to `PrinterForCommand`, receive a `nil` printer when the `outputFormat` is `wide`, and return a `HumanReadablePrinter` with the `wide` field set to `true` instead. If the output is anything else (`json`, `yaml`, etc.), it will continue to behave as normal, the only difference in implementation being that we don't version resources in `env.go` like we are currently doing, but rather receive a versioned [printer from this block](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go#L1299-L1319).

**Comment by juanvallejo** [comment]
Working example in commit https://github.com/openshift/origin/pull/11448/commits/63b6a1771b9dae52e063be38e210db164360bc81

It works with all types, but prints each object separately.

**Comment by juanvallejo** [comment]
@fabianofranz Could you take a look at https://github.com/openshift/origin/pull/11448/commits/4648b6be1b641f9fad36f3bf2d5ea39f77695a24 please?

[This checks](https://github.com/openshift/origin/pull/11448/commits/4648b6be1b641f9fad36f3bf2d5ea39f77695a24#diff-be2670f9e5fa20e06c2b074b8cc34671R497) for a `nil` printer returned from `PrinterForCommand`, and loops through each object, printing it with its own `PrinterForMapping`.

**Comment by juanvallejo** [comment]
@csrwng So I just realized that [this function](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L1262) exists, do you think it could be something that could be used by `oc set` commands, rather than the implementation I have so far? (I believe I would still have to loop through non-generic objects when calling it, as a `HumanReadablePrinter` doesn't support versioned lists :/)

**Comment by csrwng** [comment]
@juanvallejo just took a look at what you have so far and it looks fine to me. If @fabianofranz agrees, you should make the same change to the other `set` commands.

**Comment by fabianofranz** [comment]
Quick glance, but LGTM. Does it work fine for all `--output` options it adds by `AddPrinterFlags`?

**Comment by csrwng** [comment]
I tried all of them (I think) and they all seemed to work fine.

**Comment by juanvallejo** [comment]
Yeah, even `custom-columns` / key-value output formats appear to be working as expected

File: pkg/cmd/cli/cmd/set/triggers.go [hunk 1]
```diff
@@ -210,8 +210,32 @@ func (o *TriggersOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, arg
 		Flatten()
 
 	output := kcmdutil.GetFlagString(cmd, "output")
-	if len(output) != 0 {
-		o.PrintObject = func(obj runtime.Object) error { return f.PrintObject(cmd, mapper, obj, o.Out) }
+	if len(output) > 0 {
+		o.PrintObject = func(infos []*resource.Info) error {
```
Comment (juanvallejo) [inline pkg/cmd/cli/cmd/set/triggers.go]: I was not sure if there was a better way of approaching this, although this command has a function `printObjects` for cases where no `-o` option is specified, it still did not handle every outputFormat, such as `wide` without this update

**Comment by juanvallejo** [comment]
@fabianofranz @csrwng Since the following code is reused in a significant number of `oc set` subcommand updates, maybe we could turn it into a function and place it in cmdutil, WDYT?

```go
if len(v.Output) > 0 {
	v.PrintObject = func(infos []*resource.Info) error {
		printer, generic, err := kcmdutil.PrinterForCommand(cmd)
		if err != nil {
			return nil
		}
		if !generic {
			for _, info := range infos {
				mapping := info.ResourceMapping()
				printer, err := f.PrinterForMapping(cmd, mapping, false)
				if err != nil {
					return err
				}
				if err := printer.PrintObj(info.Object, v.Out); err != nil {
					return nil
				}
			}
			return nil
		}

		object, err := resource.AsVersionedObject(infos, len(infos) != 1, v.OutputVersion, kapi.Codecs.LegacyCodec(v.OutputVersion))
		if err != nil {
			return err
		}
		return printer.PrintObj(object, v.Out)
	}
}
```

It has so far been used in the `Completion` methods of `oc set buildsecret`, `oc set triggers`, `oc set volume`

**Comment by csrwng** [comment]
@juanvallejo yes, it makes sense

**Comment by fabianofranz** [comment]
SGTM

**Comment by juanvallejo** [comment]
@fabianofranz @csrwng Added a function to `cmdutil.Factory` to handle versioned printing for generic output, and custom printers from mapping for non-generic output formats, PTAL

**Comment by juanvallejo** [comment]
[test]

File: pkg/cmd/cli/cmd/set/env.go [hunk 2]
```diff
@@ -101,8 +99,8 @@ func NewCmdEnv(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
 	cmd.Flags().StringSliceVarP(&filenames, "filename", "f", filenames, "Filename, directory, or URL to file to use to edit the resource.")
 	cmd.Flags().Bool("overwrite", true, "If true, allow environment to be overwritten, otherwise reject updates that overwrite existing environment.")
 	cmd.Flags().String("resource-version", "", "If non-empty, the labels update will only succeed if this is the current resource-version for the object. Only valid when specifying a single resource.")
-	cmd.Flags().StringP("output", "o", "", "Display the changed objects instead of updating them. One of: json|yaml.")
-	cmd.Flags().String("output-version", "", "Output the changed objects with the given version (default api-version).")
+
+	kcmdutil.AddPrinterFlags(cmd)
```
Comment (juanvallejo) [inline pkg/cmd/cli/cmd/set/env.go]: @fabianofranz 
> How does the new factory method behave with --sort-by

I have tested it on almost all sub-commands (with `json`, `yaml`, `custom-columns`, `wide` output formats) so far and everything prints without issue. I have not seen if it actually sorts multiple output values as all sub-commands seem to only deal with a single resource at a time

**Comment by fabianofranz** [comment]
LGTM, squash?

**Comment by fabianofranz** [comment]
@csrwng anything else here?

**Comment by csrwng** [comment]
@fabianofranz nope, LGTM

**Comment by fabianofranz** [comment]
[merge]

**Pull request merged**